spec/p2p/v0.34/addressbook.md

```diff
@@ -0,0 +1,368 @@
+
+The [Peer Manager](./peer_manager.md) retrieves from the address book peer
+addresses to dial, so to establish outbound connections.
+It also records in the address book failed attempts to connect to a peer.
+
+The [Peer Exchange protocol](./pex-protocol.md) feds the address book with
+addresses received from peers, and retrieves from the address book random
+selection of addresses it provides to peers.
+
+Once the node is connected to a peer, it uses the address book to record
+information about the quality of the peer.
+The quality of a peer is currently restricted to the definition of
+[good](#good-peers) or [bad](#bad-peers) peers.
+It influences, in particular, the selection of candidate peers to dial.
+
+## New addresses
+
+The `AddAddress` method adds a new address to the address book.
+
+The added address is associated to a *source* address, which identifies the
+node from which the address was learned.
+
+Addresses are added to the address book in the following situations:
+
+1. When a peer address is received via PEX protocol, the source is the sender
+   of the PEX message
+1. When an inbound peer is added to the PEX reactor, in this case the peer
+   itself is set as the source of its address
+1. When the switch is instructed to dial addresses via the `DialPeersAsync`
+   method, in this case the node itself is set as the source
+
+An address contains a node ID and a network address (IP and port).
+The address book can keep multiple network addresses for the same node ID, but
+this number should be limited.
+A new address with a node ID that is already present in the address book is
+**not added** when:
+
+- The last address added with the same node ID is stored in an old bucket, so
+  it is considered a "good" address
+- There are addresses associated to the same node ID stored in
+  `maxNewBucketsPerAddress = 4` distinct buckets
+- Randomly, with a probability that increases exponentially with the number of
+  buckets in which there is an address with the same node ID.
+  So, a new address for a node ID which is already present in one bucket is
+  added with 50% of probability; if the node ID is present in two buckets, the
+  probability decreases to 25%; and if it is present in three buckets, the
+  probability is 12.5%.
+
+Addresses are stored in [buckets](#buckets).
+The bucket to which a new address is added is computed based on the address
+itself and its source address.
+An address is not added to a bucket if it is already present there.
+This should not happen with addresses with new node IDs, but it can happen with
+addresses with known node IDs.
+In fact, as the mapping of buckets for new addresses is mainly determined by
+the network group of the source address, the same address can be present in
+more than one bucket provided that it was added with different source addresses.
+
+The new address is also added to the `addrLookup` table, which stores
```

Comment:
This is too low level and in my view not really very helpful. We should try to capture spec of the address book at the more high level: what is the purpose of this structure, what are the main APIs and what invariants it should maintain? The pseudocode like description should in my view not be part of the spec but lives as a comment in the code and its purpose should be to help reader understands the code. Spec should defines higher level concepts and language we used when we talk about main p2p features.